server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1051,7 +1056,29 @@ private DeleteResult cleanupStaleIndices(Map<String, BlobContainer> foundIndices
             assert false : e;
             logger.warn(new ParameterizedMessage("[{}] Exception during cleanup of stale indices", metadata.name()), e);
         }
-        return deleteResult;
+    }
+
+    private void executeOneStaeIndexDelete(BlockingQueue<Map.Entry<String, BlobContainer>> staleIndicesToDelete, Executor executor,
+                                           GroupedActionListener<DeleteResult> listener) throws InterruptedException {
+        Map.Entry<String, BlobContainer> indexEntry  = staleIndicesToDelete.poll(0L, TimeUnit.MILLISECONDS);
+        if (indexEntry == null) {
+            listener.onResponse(DeleteResult.ZERO);
+        } else {
+            final String indexSnId = indexEntry.getKey();
+            executor.execute(ActionRunnable.supply(listener, () -> {
+                try {
+                    DeleteResult staleIndexDeleteResult = indexEntry.getValue().delete();
+                    logger.debug("[{}] Cleaned up stale index [{}]", metadata.name(), indexSnId);
+                    executeOneStaeIndexDelete( staleIndicesToDelete, executor, listener);
+                    return staleIndexDeleteResult;
+                } catch (IOException e) {
```

Comment:
I think we need to add another `catch` here now that effectively does what this block used to do (+ return the zero delete result from it as well):

https://github.com/elastic/elasticsearch/pull/64513/files#diff-b9c15e79024348d03f0af119486a84c301b54135cc309c638834c8dfd9c101b0R1052

since it's not synchronous any longer.

Reply:
Added another catch